src/compilation/options.rs

```diff
@@ -0,0 +1,87 @@
+use crate::{
+    compilation::{compile_validators, context::CompilationContext, JSONSchema, DEFAULT_SCOPE},
+    error::CompilationError,
+    resolver::Resolver,
+    schemas,
+};
+use serde_json::Value;
+use std::borrow::Cow;
+
+#[derive(Clone, Debug, Default)]
+pub struct CompilationOptions {
+    pub(crate) draft: Option<schemas::Draft>,
+}
+
+impl CompilationOptions {
+    pub(crate) fn draft(&self) -> schemas::Draft {
+        self.draft.unwrap_or_default()
+    }
+
+    /// Compile `schema` into `JSONSchema` using the currently defined options.Draft
+    pub fn compile<'a>(&self, schema: &'a Value) -> Result<JSONSchema<'a>, CompilationError> {
+        // Draft is detected in the following precedence order:
+        //   - Explicitly specified;
+        //   - $schema field in the document;
+        //   - Draft::default()
+
+        // Clone needed because we are going to store a Copy-on-Write (Cow) instance
+        // into the final JSONSchema as well as passing `self` (the instance and not
+        // the reference) would require Copy trait implementation from
+        // `CompilationOptions` which is something that we would like to avoid as
+        // options might contain heap-related objects (ie. an HashMap) and we want the
+        // memory-related operations to be explicit
+        let mut config = self.clone();
+        if self.draft.is_none() {
+            if let Some(draft) = schemas::draft_from_schema(schema) {
+                config.draft = Some(draft);
```

Comment:
I'm wondering on why a function associated to `CompilationOptions` should be use the setter instead of directly setting the field as the setter does only set the field. The end result would be the same but it would be less explicit.

But I have no strong opinion on this ... so just taking it in